.github/workflows/test-e2e.yaml

```diff
@@ -7,23 +7,55 @@ on:
 jobs:
   e2e-test:
     name: E2E Test
-    runs-on: ubuntu-latest
+    runs-on:
+      labels: ubuntu-latest-16-cores
+    env:
+      GOPATH: ${{ github.workspace }}/go
+    defaults:
+      run:
+        working-directory: ${{ env.GOPATH }}/src/github.com/kubeflow/trainer
 
     strategy:
       fail-fast: false
       matrix:
-        kubernetes-version: ["1.29.3", "1.30.0", "1.31.0"]
+        # Kubernetes versions for e2e tests on Kind cluster.
+        kubernetes-version: ["1.29.0", "1.30.0", "1.31.0"]
 
     steps:
       - name: Check out code
         uses: actions/checkout@v4
+        with:
+          path: ${{ env.GOPATH }}/src/github.com/kubeflow/trainer
+
+      - name: Setup Go
+        uses: actions/setup-go@v5
+        with:
+          go-version-file: ${{ env.GOPATH }}/src/github.com/kubeflow/trainer/go.mod
 
       - name: Setup Python
         uses: actions/setup-python@v5
         with:
-          python-version: ${{ matrix.python-version }}
+          python-version: 3.11
 
       - name: Install dependencies
         run: |
-          echo "TODO (andreyvelich): Implement E2E Tests"
-          # pip install -U './sdk'
+          echo "Install Kind"
+          go install sigs.k8s.io/kind@v0.27.0
+
```

Comment:
```suggestion
```

It would be better to manage Kind by go module. We could do the following:

1. Add the `_ sigs.k8s.io/kind` to https://github.com/kubeflow/trainer/blob/3ec8f0705f515269b5ab8744c20b9d085f50d1ce/hack/tools.go#L19-L21

2. Add make target like this one:
```makefile
KIND_VERSION ?= $(shell go list -m -f '{{.Version}}' sigs.k8s.io/kind)
KIND = $(LOCAL_BIN)/kind
.PHONY: kind
kind:
    @GOBIN=$(LOCAL_BIN) go install sig.k8s.io/kind
```

3. Add kind dependencies to test-e2e make target, `test-e2e-setup-cluster: kind`.

Reply:
Good suggestion!
Don't you see a concern to depend on [Kind Go modules](https://github.com/kubernetes-sigs/kind/blob/main/go.mod) in our Kubeflow Trainer controller manager image ?

Reply:
If we find any conflicts and concerns in the future, we could have a separate go modules only for tools.go.

Reply:
@tenzen-y I noticed this error while doing this approach
https://github.com/kubeflow/trainer/actions/runs/13663331039/job/38199243922?pr=2470#step:4:81
Any thoughts ?
```

go: downloading github.com/google/safetext v0.0.0-20220905092116-b49f7bc46da2
Error: hack/tools.go:21:2: import "sigs.k8s.io/kind" is a program, not an importable package
go test  -coverprofile cover.out
no Go files in /home/runner/work/trainer/trainer/go/src/github.com/kubeflow/trainer
make: *** [Makefile:111: test] Error 1
```

Reply:
In that case, we could add `_ "sigs.k8s.io/kind/cmd/kind/app"` to tools.go